aspnetcore/web-api/advanced/custom-formatters.md

```diff
@@ -9,112 +9,135 @@ uid: web-api/advanced/custom-formatters
+For example, suppose the action method:
 
-<a id="read-write"></a>
+* Signature returns a `Person` type.
+* It can return a `Student` or `Instructor` type that derives from `Person`. 
+
+For the formatter to handle only `Student` objects, check the type of [Object](/dotnet/api/microsoft.aspnetcore.mvc.formatters.outputformattercanwritecontext.object#Microsoft_AspNetCore_Mvc_Formatters_OutputFormatterCanWriteContext_Object) in the context object provided to the `CanWriteResult` method. When the action method returns `IActionResult`:
 
-### Override ReadRequestBodyAsync/WriteResponseBodyAsync
+*  It's not necessary to use `CanWriteResult`.
+* The `CanWriteType` method receives the runtime type.
 
-You do the actual work of deserializing or serializing in `ReadRequestBodyAsync` or `WriteResponseBodyAsync`. The highlighted lines in the following example show how to get services from the dependency injection container (you can't get them from constructor parameters).
+<a id="read-write"></a>
 
-[!code-csharp[](custom-formatters/sample/Formatters/VcardOutputFormatter.cs?name=writeresponse&highlight=3-4)]
+### Override ReadRequestBodyAsync and WriteResponseBodyAsync
 
-For an input formatter example, see the [sample app](https://github.com/dotnet/AspNetCore.Docs/tree/master/aspnetcore/web-api/advanced/custom-formatters/sample).
+Deserializing or serializing is performed in `ReadRequestBodyAsync` or `WriteResponseBodyAsync`. The following example shows how to get services from the dependency injection container. Services can't be obtained from constructor parameters.
+
+[!code-csharp[](custom-formatters/3.1sample/Formatters/VcardOutputFormatter.cs?name=writeresponse)]
 
 ## How to configure MVC to use a custom formatter
 
 To use a custom formatter, add an instance of the formatter class to the `InputFormatters` or `OutputFormatters` collection.
 
-[!code-csharp[](custom-formatters/sample/Startup.cs?name=mvcoptions&highlight=3-4)]
+::: moniker range=">= aspnetcore-3.0"
+
+[!code-csharp[](custom-formatters/3.1sample/Startup.cs?name=mvcoptions)]
+
+::: moniker-end
+
+::: moniker range="< aspnetcore-3.0"
+
+[!code-csharp[](custom-formatters/3.1sample/Startup.cs?name=mvcoptions&highlight=3-4)]
+
+::: moniker-end
 
 Formatters are evaluated in the order you insert them. The first one takes precedence.
 
-## Next steps
+## The completed `VcardInputFormatter` class
+
+The follow code shows the completed `VcardInputFormatter` class from the [sample](https://github.com/dotnet/AspNetCore.Docs/tree/master/aspnetcore/web-api/advanced/custom-formatters/3.1sample):
+
+[!code-csharp[](custom-formatters/3.1sample/Formatters/VcardInputFormatter.cs?name=snippet)]
```

Comment:
I don't think this section is needed, as the complete class is shown further up, in "How to create a custom formatter class".

Reply:
That's the `VcardOutputFormatter ` class, this is the input class.

Reply:
Ha, yes, of course! Sorry, I totally missed that. One more suggestion for that part coming up then cos of course it's staying. 😅